### DISCUSSION TITLE
Keep track of original path after moving to library?

### DISCUSSION THREAD

**Discussion by melsophos** [beetbox/beets#5392] [Q&A]
I am using Navidrome to manage my music library. I have accumulated a lot of music over the years, and I have not been very consistent with tags and filenames. I just discovered beets and would like to refresh my library. The problem is that Navidrome identifies a file by its path: so if a file is renamed/moved, it deletes everything about it in the database. Hence, to avoid loosing my statistics, I need to keep track of tracks moved by beets. Then I can use scripts (like [this one](https://github.com/kgarner7/navidrome-migrate)) to update Navidrome database.

I am wondering if there is any way to keep track of how beets maps files in the original folder and in the library after using `move/copy`? The `import: log` does not display much information. The text displayed while running is painful to parse.

The same problem discussed in discussion #5088, but the user decides to not move the files, which I don't find satisfactory. Proposals to log more details have been given in  #141 and #1392, but they have not been implemented; since I don't need so many features (just filename updates), I was hoping to get away with something simpler.

**Comment by arogl** [comment]
Have a look through some of these for inspiration on how you might be able to create a flexible field in a plugin

You might have access to original paths and filenames in the plugin

[misc beets plugins](https://github.com/kergoth/beets-kergoth)

**Comment by RollingStar** [comment]
I wanted this long ago. https://discourse.beets.io/t/should-we-start-recording-the-original-path-and-filename-on-import/1494/7

Looks like dirfields does it. https://github.com/JOJ0/beets-dirfields

**Reply by melsophos** [reply]
I had indeed found the discussion on Discourse, the idea of storing the path in a tag was good. I had not found the plugin you are linking, thanks for that.

At the end, I implemented a quick and dirty solution of symlinking and using a Python script to list all symlinks to update the database:

````python
#!/usr/bin/python

"""
Script to list path changes in a music folder after using beets.

This is useful to update paths in Navidrome database. The process works as follows.
1. Run `beet` using `link: yes`. This creates symlinks pointing to the original files.
2. Run the current script to make a list associating the new path with the old path, which is
   stored in a text file. This also replaces the symlink by the file, deleting the old file.
3. Run the bash script below to update Navidrome database. This requires the following script:
    https://gist.github.com/bagaag/3d64e3349b6ed3bfd6e01813222db055

Update Navidrome:
```
cd /docker
docker compose stop navidrome

cd /services/navidrome/data
cp navidrome.db navidrome.db.back

# run the current script now

while read -r line; do
	python change_path.py "${line%;;*}" "${line#*;;}"
done < updated_files.txt
```
"""

import os
import shutil


# update if needed
LOCAL_MUSIC_FOLDER = "~/Music"
NAVIDROME_MUSIC_FOLDER = "/music"


def list_symlinks():

    pairs = []

    for dirpath, _, filenames in os.walk(LOCAL_MUSIC_FOLDER):
        for filename in filenames:
            path = os.path.join(dirpath, filename)

            if os.path.islink(path):
                pair = replace_symlink(path)
                pairs.append(pair)

    with open("updated_files.txt", "w") as f:
        f.writelines(pairs)

    return


def replace_symlink(path):
    source = os.readlink(path)
    os.remove(path)
    shutil.move(source, path)

    return f"{source};;{path}\n".replace(LOCAL_MUSIC_FOLDER, NAVIDROME_MUSIC_FOLDER)

list_symlinks()
```

Next, delete empty folders (you may need to delete leftover files like covers first):
```
find ~/Music/ -type d -empty -delete
```

**Comment by kergoth** [comment]
https://github.com/beetbox/beets/pull/4748 is likely of interest to you.
